File: web/src/lib/filters.ts
/**
 * TypeScript mirror of `api/filters.py::DashboardFilters` plus the
 * URL <-> filter-state conversions the client needs. Field names and
 * defaults must match the Pydantic model exactly — see that module's
 * docstring for the invariants these filters carry (two-frame filtering,
 * month-granularity periods, etc).
 */

import type { FilterOptions } from './types';

export type PeriodPreset =
  'last_30_days' | 'current_month' | 'last_3_months' | 'last_6_months' | 'ytd' | 'all_time' | 'custom';

const VALID_PERIODS: ReadonlySet<string> = new Set<PeriodPreset>([
  'last_30_days',
  'current_month',
  'last_3_months',
  'last_6_months',
  'ytd',
  'all_time',
  'custom',
]);

// Mirrors api/filters.py::DEFAULT_PERIOD. Deliberately wider than Streamlit's
// last_30_days — see that module's comment.
export const DEFAULT_PERIOD: PeriodPreset = 'last_3_months';

export const PERIOD_LABELS: Record<PeriodPreset, string> = {
  last_30_days: 'Last 30 days',
  current_month: 'This month',
  last_3_months: 'Last 3 months',
  last_6_months: 'Last 6 months',
  ytd: 'Year to date',
  all_time: 'All time',
  custom: 'Custom',
};

export interface DashboardFilters {
  period: PeriodPreset;
  /** Explicit `YYYY-MM` keys; only consulted when period === 'custom'. */
  months: string[] | null;
  owners: string[] | null;
  categories: string[] | null;
  accounts: string[] | null;
  amount_min: number | null;
  amount_max: number | null;
  search: string | null;
  outliers_only: boolean;
  duplicates_only: boolean;
}

export const DEFAULT_FILTERS: DashboardFilters = {
  period: DEFAULT_PERIOD,
  months: null,
  owners: null,
  categories: null,
  accounts: null,
  amount_min: null,
  amount_max: null,
  search: null,
  outliers_only: false,
  duplicates_only: false,
};

// ---------------------------------------------------------------------------
// Filters <-> URLSearchParams
// ---------------------------------------------------------------------------

function appendArray(params: URLSearchParams, key: string, values: string[] | null | undefined): void {
  if (!values || values.length === 0) return;
  for (const value of values) params.append(key, value);
}

/**
 * Build the query string FastAPI expects: repeated keys for array params
 * (`?owners=A&owners=B`), and anything already at its default omitted
 * entirely so a fully-default filter set produces an empty/clean URL.
 */
export function toSearchParams(filters: DashboardFilters): URLSearchParams {
  const params = new URLSearchParams();

  if (filters.period !== DEFAULT_PERIOD) {
    params.set('period', filters.period);
  }
  appendArray(params, 'months', filters.months);
  appendArray(params, 'owners', filters.owners);
  appendArray(params, 'categories', filters.categories);
  appendArray(params, 'accounts', filters.accounts);

  if (filters.amount_min !== null && filters.amount_min !== undefined) {
    params.set('amount_min', String(filters.amount_min));
  }
  if (filters.amount_max !== null && filters.amount_max !== undefined) {
    params.set('amount_max', String(filters.amount_max));
  }
  if (filters.search) {
    params.set('search', filters.search);
  }
  if (filters.outliers_only) {
    params.set('outliers_only', 'true');
  }
  if (filters.duplicates_only) {
    params.set('duplicates_only', 'true');
  }

  return params;
}

function parsePeriod(raw: string | null): PeriodPreset {
  return raw && VALID_PERIODS.has(raw) ? (raw as PeriodPreset) : DEFAULT_PERIOD;
}

function parseArray(params: URLSearchParams, key: string): string[] | null {
  const values = params.getAll(key);
  return values.length > 0 ? values : null;
}

function parseNumber(raw: string | null): number | null {
  if (raw === null || raw === '') return null;
  const value = Number(raw);
  return Number.isFinite(value) ? value : null;
}

function parseBool(raw: string | null): boolean {
  return raw === 'true' || raw === '1';
}

/**
 * Parse a (possibly hand-edited) URL's query string back into filter state.
 * Never throws: an unknown period falls back to the default, a non-numeric
 * amount becomes `null`, and any param can simply be absent.
 */
export function fromSearchParams(params: URLSearchParams): DashboardFilters {
  return {
    period: parsePeriod(params.get('period')),
    months: parseArray(params, 'months'),
    owners: parseArray(params, 'owners'),
    categories: parseArray(params, 'categories'),
    accounts: parseArray(params, 'accounts'),
    amount_min: parseNumber(params.get('amount_min')),
    amount_max: parseNumber(params.get('amount_max')),
    search: params.get('search') || null,
    outliers_only: parseBool(params.get('outliers_only')),
    duplicates_only: parseBool(params.get('duplicates_only')),
  };
}

// ---------------------------------------------------------------------------
// Chips / active-filter summary, for the filter bar UI
// ---------------------------------------------------------------------------

export interface FilterChip {
  /** Stable key for React lists; also identifies which value a chip removes. */
  id: string;
  label: string;
  /** Pure: returns the next filter state with this one thing removed. */
  remove: (filters: DashboardFilters) => DashboardFilters;
}

function monthLabel(key: string, options?: FilterOptions): string {
  return options?.months.find((month) => month.key === key)?.label ?? key;
}

function withoutArrayValue(values: string[] | null, value: string): string[] | null {
  if (!values) return null;
  const next = values.filter((entry) => entry !== value);
  return next.length > 0 ? next : null;
}

/**
 * One chip per active filter *value* (so selecting 3 owners produces 3
 * removable chips), plus one chip for the period when it isn't the default.
 * `options` is optional so chips can render before `/filter-options` has
 * loaded — months just fall back to their raw `YYYY-MM` key.
 */
export function activeFilterChips(filters: DashboardFilters, options?: FilterOptions): FilterChip[] {
  const chips: FilterChip[] = [];

  if (filters.period !== DEFAULT_PERIOD) {
    chips.push({
      id: 'period',
      label: `Period: ${PERIOD_LABELS[filters.period]}`,
      remove: (f) => ({ ...f, period: DEFAULT_PERIOD, months: null }),
    });
  }

  if (filters.period === 'custom' && filters.months) {
    for (const month of filters.months) {
      chips.push({
        id: `months:${month}`,
        label: monthLabel(month, options),
        remove: (f) => ({ ...f, months: withoutArrayValue(f.months, month) }),
      });
    }
  }

  for (const owner of filters.owners ?? []) {
    chips.push({
      id: `owners:${owner}`,
      label: owner,
      remove: (f) => ({ ...f, owners: withoutArrayValue(f.owners, owner) }),
    });
  }

  for (const category of filters.categories ?? []) {
    chips.push({
      id: `categories:${category}`,
      label: category,
      remove: (f) => ({ ...f, categories: withoutArrayValue(f.categories, category) }),
    });
  }

  for (const account of filters.accounts ?? []) {
    chips.push({
      id: `accounts:${account}`,
      label: account,
      remove: (f) => ({ ...f, accounts: withoutArrayValue(f.accounts, account) }),
    });
  }

  if (filters.amount_min !== null || filters.amount_max !== null) {
    const parts: string[] = [];
    if (filters.amount_min !== null) parts.push(`≥ $${filters.amount_min}`);
    if (filters.amount_max !== null) parts.push(`≤ $${filters.amount_max}`);
    chips.push({
      id: 'amount',
      label: `Amount ${parts.join(', ')}`,
      remove: (f) => ({ ...f, amount_min: null, amount_max: null }),
    });
  }

  if (filters.search) {
    chips.push({
      id: 'search',
      label: `Search: "${filters.search}"`,
      remove: (f) => ({ ...f, search: null }),
    });
  }

  if (filters.outliers_only) {
    chips.push({
      id: 'outliers_only',
      label: 'Flagged only',
      remove: (f) => ({ ...f, outliers_only: false }),
    });
  }

  if (filters.duplicates_only) {
    chips.push({
      id: 'duplicates_only',
      label: 'Possible duplicates only',
      remove: (f) => ({ ...f, duplicates_only: false }),
    });
  }

  return chips;
}

/**
 * Number of active filter *axes* (not values) — what a collapsed
 * "Filters (N)" button should show. Custom months count as one axis
 * alongside (not on top of) the period axis, since they're one control.
 */
export function countActiveFilters(filters: DashboardFilters): number {
  let count = 0;
  if (filters.period !== DEFAULT_PERIOD) count += 1;
  if (filters.period === 'custom' && filters.months && filters.months.length > 0) count += 1;
  if (filters.owners && filters.owners.length > 0) count += 1;
  if (filters.categories && filters.categories.length > 0) count += 1;
  if (filters.accounts && filters.accounts.length > 0) count += 1;
  if (filters.amount_min !== null || filters.amount_max !== null) count += 1;
  if (filters.search) count += 1;
  if (filters.outliers_only) count += 1;
  if (filters.duplicates_only) count += 1;
  return count;
}

File: api/filters.py
"""Period/owner/category/... filtering for the dashboard read endpoints.

This is a **port**, not an extraction, of `app/dashboard.py::_build_sidebar_filters`
(lines 710-812). `app/dashboard.py` is frozen as of PLAN.md Phase 15, so the masking
logic is reimplemented here rather than lifted into a module both callers share. That
buys a real drift risk, which `tests/test_api_filters.py` exists to catch: every
non-obvious behaviour below is pinned by a test naming the Streamlit line it mirrors.

Three invariants carried over deliberately — each is load-bearing and each is easy to
"clean up" into a bug:

1. **Two frames, not one.** `apply_filters` returns `(filtered, all_time)`. `all_time`
   has every non-date filter applied but ignores the period, because trend charts and
   the emergency-fund metric must not collapse to a single point when the user picks a
   30-day window. Mirrors `_build_sidebar_filters`'s two return values.

2. **Filtering happens strictly downstream of enrichment.** `prepare_transactions` runs
   once over the complete dataset so internal-transfer pair matching sees both legs of a
   transfer even when one leg's account is filtered out. Never filter before enriching.

3. **The period selects whole MONTHS, not a date range.** `_preset_month_keys` resolves a
   preset to a date window, then keeps every month that window *touches* — so "last 30
   days" spanning a month boundary includes both months in full. This looks like a bug
   and is not: every aggregate downstream (budgets, month-over-month, the monthly trend
   lines) groups by month, and admitting half a month would render a partial month's
   spend as if it were a whole one. Month granularity in, month granularity out.

Anchoring also follows Streamlit: windows are measured back from the **latest transaction
date**, not from today. With a daily pipeline those coincide; when ingestion has stalled,
anchoring to `today` would silently return an empty dashboard instead of the most recent
data that exists.
"""

from __future__ import annotations

from datetime import date
from typing import Annotated, Literal

import pandas as pd
from fastapi import Depends, Query
from pydantic import BaseModel, Field

PeriodPreset = Literal[
    "last_30_days",
    "current_month",
    "last_3_months",
    "last_6_months",
    "ytd",
    "all_time",
    "custom",
]

# Streamlit defaults to last_30_days. The React dashboard defaults wider: with a 30-day
# window every "average monthly" figure averages exactly one month, which is the metric
# the user reported as misleading in the first place. See PLAN.md Phase 15, Fix 9.
DEFAULT_PERIOD: PeriodPreset = "last_3_months"


class DashboardFilters(BaseModel):
    """The 10 Streamlit sidebar filters, as query parameters."""

    period: PeriodPreset = DEFAULT_PERIOD
    months: list[str] | None = Field(
        default=None, description="Explicit `YYYY-MM` keys; only consulted when period='custom'."
    )
    owners: list[str] | None = None
    categories: list[str] | None = None
    accounts: list[str] | None = None
    amount_min: float | None = None
    amount_max: float | None = None
    search: str | None = None
    outliers_only: bool = False
    duplicates_only: bool = False
    date_from: date | None = Field(
        default=None,
        description=(
            "Inclusive day-level lower bound, narrowing WITHIN whatever `period`/`months` already "
            "selects -- NOT a replacement for month-level filtering (the module's month-granularity "
            "invariant above is unchanged). Callers must ensure `period`/`months` already cover this "
            "range; `date_from`/`date_to` narrow further, they never widen. Added for a Cash Flow "
            "dashboard drill-down: jumping to a filtered Transactions view for exactly one week or one "
            "rolling-30-day window needs day precision the month-granularity filters can't express alone."
        ),
    )
    date_to: date | None = Field(
        default=None, description="Inclusive day-level upper bound. See `date_from`."
    )

    def is_default(self) -> bool:
        """True when nothing narrows the data — lets callers skip work and lets the UI
        decide whether to show a "clear filters" affordance."""
        return (
            self.period == DEFAULT_PERIOD
            and not self.months
            and not self.owners
            and not self.categories
            and not self.accounts
            and self.amount_min is None
            and self.amount_max is None
            and not self.search
            and not self.outliers_only
            and not self.duplicates_only
            and self.date_from is None
            and self.date_to is None
        )


def dashboard_filters(
    period: Annotated[PeriodPreset, Query()] = DEFAULT_PERIOD,
    months: Annotated[list[str] | None, Query()] = None,
    owners: Annotated[list[str] | None, Query()] = None,
    categories: Annotated[list[str] | None, Query()] = None,
    accounts: Annotated[list[str] | None, Query()] = None,
    amount_min: Annotated[float | None, Query()] = None,
    amount_max: Annotated[float | None, Query()] = None,
    search: Annotated[str | None, Query()] = None,
    outliers_only: Annotated[bool, Query()] = False,
    duplicates_only: Annotated[bool, Query()] = False,
    date_from: Annotated[date | None, Query()] = None,
    date_to: Annotated[date | None, Query()] = None,
) -> DashboardFilters:
    """FastAPI query-param dependency. Repeated params (`?owners=A&owners=B`) arrive as
    lists; omitting one entirely means "no constraint", which is not the same as an empty
    list, so the defaults are `None` rather than `[]`."""
    return DashboardFilters(
        period=period,
        months=months,
        owners=owners,
        categories=categories,
        accounts=accounts,
        amount_min=amount_min,
        amount_max=amount_max,
        search=search,
        outliers_only=outliers_only,
        duplicates_only=duplicates_only,
        date_from=date_from,
        date_to=date_to,
    )


FiltersDep = Annotated[DashboardFilters, Depends(dashboard_filters)]


def preset_month_keys(preset: PeriodPreset, df: pd.DataFrame) -> set[str]:
    """Month keys a quick-range preset touches, anchored to the latest transaction date.

    Port of `app/dashboard.py::_preset_month_keys` (lines 689-707). Reuses the `month`
    column `_enrich_transactions` already computes rather than recomputing Streamlit's
    private `_month_key` — same formula (`date.dt.to_period("M").astype(str)`).
    """
    if df.empty:
        return set()

    max_date = df["date"].max()
    if preset == "all_time":
        return set(df["month"].unique())

    if preset == "last_30_days":
        start = max_date - pd.Timedelta(days=29)
    elif preset == "current_month":
        start = max_date.replace(day=1)
    elif preset == "last_3_months":
        start = max_date - pd.DateOffset(months=3)
    elif preset == "last_6_months":
        start = max_date - pd.DateOffset(months=6)
    elif preset == "ytd":
        start = max_date.replace(month=1, day=1)
    else:
        start = df["date"].min()

    window = df[(df["date"] >= start) & (df["date"] <= max_date)]
    return set(window["month"].unique())


def _non_date_mask(df: pd.DataFrame, filters: DashboardFilters) -> pd.Series:
    """Every filter except the period. Port of `app/dashboard.py` lines 782-805."""
    mask = pd.Series(True, index=df.index)

    if filters.owners:
        mask &= df["owner_name"].isin(filters.owners)
    if filters.categories:
        mask &= df["category"].isin(filters.categories)
    if filters.accounts:
        mask &= df["account_name"].isin(filters.accounts)

    if filters.amount_min is not None or filters.amount_max is not None:
        abs_amount = df["amount"].abs()
        lo = filters.amount_min
        hi = filters.amount_max
        # Tolerate an inverted range rather than silently returning zero rows —
        # `app/dashboard.py:768` makes the same allowance.
        if lo is not None and hi is not None and lo > hi:
            lo, hi = hi, lo
        if lo is not None:
            mask &= abs_amount >= lo
        if hi is not None:
            mask &= abs_amount <= hi

    if filters.search:
        # `regex=False` is a deliberate divergence from Streamlit, which leaves
        # `str.contains` in its default regex mode. Over HTTP that turns a description
        # search for "(" into a 500; a literal substring match is also what someone
        # typing into a search box actually means.
        mask &= df["description"].str.contains(filters.search, case=False, na=False, regex=False)

    if filters.outliers_only:
        mask &= df["is_outlier"].fillna(False).astype(bool)

    if filters.duplicates_only:
        # Grouped on `account_key`, NOT `account_name`: two distinct accounts can share a
        # display name, and collapsing them would invent duplicates that don't exist.
        # Mirrors `app/dashboard.py:788-792`.
        group_sizes = df.groupby(["account_key", "date", "description", "amount"], dropna=False)[
            "amount"
        ].transform("size")
        mask &= group_sizes > 1

    return mask


def apply_filters(df: pd.DataFrame, filters: DashboardFilters) -> tuple[pd.DataFrame, pd.DataFrame]:
    """Return `(filtered, all_time)`.

    `filtered` has every filter applied. `all_time` has everything except the period, and
    is what trend charts consume — see invariant 1 in the module docstring.

    `df` must already be enriched (`prepare_transactions`), never raw.
    """
    if df.empty:
        return df, df

    non_date = _non_date_mask(df, filters)

    if filters.period == "custom":
        selected_months = set(filters.months or [])
    else:
        selected_months = preset_month_keys(filters.period, df)

    date_mask = df["month"].isin(selected_months)
    # Day-level narrowing WITHIN the month selection above -- see `date_from`/`date_to`'s
    # docstrings. These never widen the window: a day outside the selected months stays
    # excluded even if it falls inside [date_from, date_to].
    if filters.date_from is not None:
        date_mask &= df["date"] >= pd.Timestamp(filters.date_from)
    if filters.date_to is not None:
        date_mask &= df["date"] <= pd.Timestamp(filters.date_to)
    return df[non_date & date_mask], df[non_date]


def build_filter_options(df: pd.DataFrame) -> dict:
    """Everything the filter UI needs to populate its controls, derived from the
    unfiltered frame so the option lists don't shrink as the user narrows the view."""
    if df.empty:
        return {
            "owners": [],
            "categories": [],
            "accounts": [],
            "months": [],
            "amount_min": 0.0,
            "amount_max": 0.0,
        }

    month_labels = (
        df[["month"]]
        .drop_duplicates()
        .assign(label=lambda x: pd.to_datetime(x["month"] + "-01").dt.strftime("%B %Y"))
        .sort_values("month")
    )
    abs_amounts = df["amount"].abs()
    return {
        "owners": sorted(df["owner_name"].dropna().unique().tolist()),
        "categories": sorted(df["category"].dropna().unique().tolist()),
        "accounts": sorted(df["account_name"].dropna().unique().tolist()),
        "months": [{"key": row["month"], "label": row["label"]} for _, row in month_labels.iterrows()],
        "amount_min": float(abs_amounts.min()),
        "amount_max": float(abs_amounts.max()),
    }
